Register consecutive priorities in batch

Dyanngg opened the pull request:
This PR aims to improve the efficiency of registering Antrea-native policy priorities in the OpenFlow priority space. Fundamental idea is to take advantage of the fact that, when the agent receives internal NP created for those policies, it already knows the number of absolute consecutive Priorities that needs to be registered. Those Priorities will have the same Tier and Policy priority, and consecutive rule priorities as they correspond to the 1st, 2nd, and etc. rules of the ingress/egress rule of that policy.

Thus, when registering priorities, we can break these priorities into zones (each zone of same Tier and Policy priority), and register each zone in batch. This eliminates the problematic scenario where asynchronously registering a series of priorities cause existing priorities to be reassigned each time.

The priorityAssigner module is refactored to accommodate this change. The following illustrations demonstrates the new algorithm that it uses to insert consecutive priorities. For each diagram, the first row depicts the current ofPriority space, and where the initial heuristic function puts in the inserting priorities. 

**Case 1** The spaces computed by the heuristic function is unoccupied, AND the priorities immediately lower than those being inserted, if exists, are registered below these ofPriorities, AND the priorities immediately higher than those being inserted, if exists, are registered above these ofPriorities. In this case, we simply assign the ofPriorities returned by the heuristic function.
![image](https://user-images.githubusercontent.com/11527024/94870242-f2cb0c80-03fb-11eb-84a9-1db87d71c860.png)

**Case 2** The ofPriorities returned by the heuristic function overlaps with existing priorities/are out of place. If the priorities to be registered overlaps with lower priorities/are lower than the lower priorities, and the gap between lowerBound for insertion and upperBound for insertion is very large, then we insert these priorities above the lowerBound, offsetted by a constant zoneOffset. This offset gives buffer in case priorities are again created in between those zone. Vice versa for priorities overlaps with/higher than higher priorities.
![image](https://user-images.githubusercontent.com/11527024/94870569-af24d280-03fc-11eb-89a6-87733993a742.png)

**Case 3** The ofPriorities returned by the heuristic function overlaps with existing priorities/are out of place, and the gap between lowerBound for insertion and upperBound for insertion is tight. In that case, we simply register the priorities in the middle of the gap.
![image](https://user-images.githubusercontent.com/11527024/94870623-dc718080-03fc-11eb-8f83-ddb7995e3a33.png)

**Case 4** The gap between lowerBound for insertion and upperBound for insertion is not big enough for insertion. In this case, we push lower priorities and/or higher priorities towards each side in attempt to make enough room for the registering priorities, while in the meantime try to reassign as few existing priorities as possible. For example, in the following diagram, the yellow Priorities will be kept where they were to save reassignments.
![image](https://user-images.githubusercontent.com/11527024/95159664-72860d80-0753-11eb-8860-66049874da31.png)

antoninbas commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +98,220 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
 		}
-	default:
-		return nil, map[uint16]uint16{}, nil, fmt.Errorf("no available Openflow priority left to insert priority %v", p)
+		gap++
```
why does the gap only increase by 1 instead of `costSiftDown` / `costSiftUp`?

Dyanngg replied:
This is indeed an issue... Thanks for spotting it! Because of this, one of the UT test cases is not expecting the correct outcome and I missed it. Refactored the reassign logic and updated the UT.

antoninbas commented on pkg/agent/types/networkpolicy.go:
```diff
@@ -81,6 +81,19 @@ func (p *Priority) Equals(p2 Priority) bool {
+	return p.PolicyPriority == p2.PolicyPriority && p.TierPriority == p2.TierPriority
+}
+
+// IsConsecutive returns if two Priorties are immediately next to each other.
```
missing "true" word

antoninbas commented on pkg/agent/types/networkpolicy.go:
```diff
@@ -81,6 +81,19 @@ func (p *Priority) Equals(p2 Priority) bool {
 	return p.TierPriority == p2.TierPriority && p.PolicyPriority == p2.PolicyPriority && p.RulePriority == p2.RulePriority
 }
 
+// InSamePriorityZone returns if two Priorities are of the same Tier and same priority at policy level.
```
missing "true" word

antoninbas commented on pkg/agent/controller/networkpolicy/priority_test.go:
```diff
@@ -72,275 +73,221 @@ func TestUpdatePriorityAssignment(t *testing.T) {
-			p1131,
-			map[types.Priority]uint16{
-				p1120: 10000, p1121: 9999, p1130: 9998, p1140: 9997, p1141: 9996, p1160: 9995},
+			"clutch-priorities",
```
what's "clutch" priorities?

Dyanngg replied:
rephrased.

antoninbas reviewed (commented):
Just want to make sure that your unit tests cover all the cases described in the PR description (thanks for the comprehensive description BTW!) ?

Dyanngg commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +98,220 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
-func (pa *priorityAssigner) getLastVacantOFPriority(ofPriority uint16) *uint16 {
+// starting from, but not including, the input ofPriority. It also returns the distance between input
+// ofPriority and last vacant ofPriority, as well as all the registered Priorities in between.
+func (pa *priorityAssigner) getLastVacantOFPriority(ofPriority uint16) (*uint16, uint16, types.ByPriority) {
```
Refactored this part.

Dyanngg commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +98,220 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+	// gap keeps track of the vacant ofPriority thus far between lowerBound and upperBound.
+	gap := upperBound - lowerBound - 1
+	target := uint16(len(prioritiesToRegister))
+	// siftedPrioritiesLow and siftedPrioritiesHigh keeps track of Priorities that needs to be reassigned,
```
Done.

Dyanngg commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -24,18 +24,36 @@ import (
+	MaxUint16                = ^uint16(0)
 	PolicyBottomPriority     = uint16(100)
 	PolicyTopPriority        = uint16(65000)
+	zoneOffset               = uint16(5)
```
Added in the comments where this zoneOffset is used in code.

Dyanngg commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +98,220 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+		break
+	// there are some overlaps between upper/lowerBound and insertionPointLow/High, and the window between
+	// upper/lowerBound is large. Assign Priorities by offsetting the upper/lowerBound, depending on where
+	// the overlap is. The rational is that overlapped Priorities would most likely to be more adjacent to
+	// the registering Priorities.
```
Rephrased the comments.

Dyanngg commented:
> Just want to make sure that your unit tests cover all the cases described in the PR description (thanks for the comprehensive description BTW!) ?

I have updated UT and I believe it now covers all the cases mentioned in the PR description. Also, I realized that the original PR description for case 4 is inaccurate, I have updated it accordingly as well.

Dyanngg commented:
/test-all

antoninbas commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+	// To reach the target number of slots to be added into the gap, Priorities needs to be sifted upwards or
+	// downwards (or both), and empty slots from lower and higher ofPriority space will be swapped into the gap.
+	// costMap maintains the costs and reassign boundaries for each combination of lower empty slots used and
+	// higher empty slots used, which makes up to the target. For example, if the target is 2, the maps stores
```
s/which makes up to the target/with the sum making up the target

antoninbas commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+	for i := target; i >= 0; i-- {
+		if cost, exists := costMap[i]; exists && cost.cost < minCost {
+			// make sure that the reassign range makes up for all Priorities to be registered.
+			if int(cost.upperBound-cost.lowerBound)+1 == numNewPriorities+cost.cost {
```
isn't this guaranteed by your algorithm above? If yes, wouldn't it make sense to log something or return an error if this doesn't hold?

Dyanngg replied:
No it is not guaranteed by the algorithm above and there should not be an error thrown. Consider a case where we need to find 3 empty spots below lowerBound and/or above upperBound. In the first pass, the algorithm tries to fill the table on where the lowerBound would be at if we were to use 1, 2 and 3 lower empty spots. For 1 and 2, when a cost is found, the upperBound will be temporarily set to the current upperBound which has not shifted at all. Those costs however only makes sense if in the second pass, the algorithm is able to find cost of getting 2 and 1 empty spots from above, in which case the cost and new upperBound will be updated. 
In other words, when `int(cost.upperBound-cost.lowerBound)+1 != numNewPriorities+cost.cost`, it simply means that this combination of (use x empty slots below, use y empty slots above) is not feasible based on the current map, but it's not an error.

antoninbas commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
-			return &i
+// reassignBoundaryPriorities reassigns Priorities from lowerBound / upperBound or both, to make room for
+// new Priorities to be registered. It also records all the priority updates due to the reassignment in the
+// map updates which is passed to it as parameter.
```
s/map updates/updates map

or

s/map updates/map of updates

Dyanngg replied:
Done.

antoninbas commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+	if err != nil {
+		return err
+	}
+	// siftedPrioritiesLow and siftedPrioritiesHigh keeps track of Priorities that need to be reassigned,
```
s/keeps track/keep track

antoninbas commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+// input Priorities that are not yet registered. It also returns the ofPriority updates if there are reassignments,
+// as well as a revert function that can undo the registration if any error occurred in data plane.
+func (pa *priorityAssigner) RegisterPriorities(priorities []types.Priority) (map[uint16]uint16, func(), error) {
+	// create a zero-length slice with the same underlying array
+	prioritiesToRegister := priorities[:0]
```
just want to make sure that you do intend to overwrite the input `priorities` slice here? you may want to mention this in the function comment since it can be a bit surprising when calling the function

Dyanngg replied:
Yes. Updated the comment to reflect that.

antoninbas commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
-		return fmt.Errorf("cannot batch register priorities with a non empty priority map")
+// insertConsecutivePriorities inserts a list of consecutive Priorities into the ofPriority space.
+// It first identifies the lower and upper bound for insertion, by obtaining the ofPriorities of
+// registered Priority that is immediately lower and higher than the inserting Priorities. It then
```
do you mean "by obtaining the ofPriorities of the registered Priorities surrounding (immediately lower and higher) the Priorities being inserted"?

Dyanngg replied:
Thanks for the suggestion. Done.

antoninbas commented on pkg/agent/controller/networkpolicy/priority.go:
```diff
@@ -79,272 +107,246 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+	// ofPriorities provided by the heuristic function are good.
+	case insertionPointLow > lowerBound && insertionPointHigh < upperBound:
+		break
+	// ofPriorities returned by the heuristic function overlaps with existing Priorities/are out of place.
```
s/overlaps/overlap

antoninbas reviewed (commented):
LGTM but we should get another pair of eyes on this

Dyanngg commented:
/test-all

Dyanngg commented:
/test-all
